tests/Queries/QueryBuilderFactoryTest.php

```diff
@@ -0,0 +1,132 @@
+        \Doctrine\ORM\Tools\Setup::createAnnotationMetadataConfiguration(
+            array(__DIR__."/src"),
+            true
+        ));
+    }
+
+    public function testCanFilterThanksToOperators()
+    {
+        $this->manager = $this
+            ->getMockBuilder('Doctrine\ORM\EntityManager')
+            ->disableOriginalConstructor()
+            ->getMock();
+
+        $queryBuilderFactory = new QueryBuilderFactory($this->manager);
+
+        $this->assertEquals(
+            array_keys(Operators::getAll()),
+            $queryBuilderFactory->getAvailableFilters()
+        );
+    }
+
+    /**
+     * @expectedException \RuntimeException
+     */
+    public function testThrowExceptionWheneverUsedWithoutFields()
+    {
+        $this->manager = $this
+            ->getMockBuilder('Doctrine\ORM\EntityManager')
+            ->disableOriginalConstructor()
+            ->getMock();
+
+        $queryBuilderFactory = new QueryBuilderFactory($this->manager);
+        $queryBuilderFactory->ensureFieldsDefinedPublic();
+    }
+
+    public function testBuildQueryHandly()
+    {
+        $queryBuilderFactory = new QueryBuilderFactory($this->manager);
+        $queryBuilderFactory->setFields([ 'id' ]);
+        $queryBuilderFactory->setFilters([ 'id|eq' => 33 ]);
+        $queryBuilderFactory->createQueryBuilder(MySimpleEntity::class, 'e');
+        $queryBuilderFactory->filter();
+
+        $this->assertEquals(
+            "SELECT m0_.id AS id_0 FROM MySimpleEntity m0_ WHERE m0_.id = ?",
+            $queryBuilderFactory->getQueryBuilder()->getQuery()->getSql()
+        );
+
+        $this->assertEquals(
+            "SELECT e FROM Mado\QueryBundle\Tests\Objects\MySimpleEntity e WHERE e.id = :field_id",
+            $queryBuilderFactory->getQueryBuilder()->getQuery()->getDql()
+        );
+    }
+
+    public function testOneToMany()
+    {
+        $queryBuilderFactory = new QueryBuilderFactory($this->manager);
+        $queryBuilderFactory->setFields([ 'id' ]);
+        $queryBuilderFactory->setRel([ 'group' ]);
+        $queryBuilderFactory->setFilters([
+            '_embedded.group.name|contains|1' => 'ad',
+            '_embedded.group.name|contains|2' => 'ns',
+            '_embedded.group.name|contains|3' => 'dm',
+            '_embedded.group.name|contains|4' => 'mi',
+        ]);
+        $queryBuilderFactory->createQueryBuilder(User::class, 'e');
+        $queryBuilderFactory->filter();
+
+        $this->assertEquals(
+            "SELECT" .
+            " u0_.id AS id_0," .
+            " u0_.username AS username_1," .
+            " u0_.group_id AS group_id_2 " .
+            "FROM User u0_ " .
+            "INNER JOIN Group g1_ ON u0_.group_id = g1_.id " .
+            "WHERE g1_.name LIKE ? " .
+            "AND g1_.name LIKE ? " .
+            "AND g1_.name LIKE ? " .
+            "AND g1_.name LIKE ?",
+            $queryBuilderFactory->getQueryBuilder()->getQuery()->getSql()
+        );
+    }
+}
+
+/** @Entity() */
```

Comment:
Is better to move this into separate files?

Reply:
What do you think about this folding structure:

tests/
└── resources
 |  └── ResourceName.php
└── unitary
 |  └── SomeUnitaryTest.php
└── integration
    └── SomeIntegrationTest.php

???